Fallback logging in ModelAssembler should fulfill MessageFormat constraints

Let's make sure issue is not already fixed in latest builds first.

- [x] I verified I can reproduce this issue against [latest Integration Build of Eclipse SDK](https://download.eclipse.org/eclipse/downloads/)

## Steps to reproduce

From a fresh installation and clean workspace:

* Launch a custom e4 application

I tried

* Shutdown the application and the OSGi environment

I expected: readable error messages

But got: 

```
tatus ERROR: org.eclipse.ui code=0 Unhandled event loop exception java.lang.IllegalArgumentException: can't parse argument number: 
java.lang.IllegalArgumentException: can't parse argument number: 
	at java.base/java.text.MessageFormat.makeFormat(MessageFormat.java:1485)
	at java.base/java.text.MessageFormat.applyPattern(MessageFormat.java:512)
	at java.base/java.text.MessageFormat.<init>(MessageFormat.java:383)
	at java.base/java.text.MessageFormat.format(MessageFormat.java:883)
	at org.eclipse.e4.ui.internal.workbench.ModelAssembler.log(ModelAssembler.java:823)
	at org.eclipse.e4.ui.internal.workbench.ModelAssembler.warn(ModelAssembler.java:806)
	at org.eclipse.e4.ui.internal.workbench.ModelAssembler.lambda$1(ModelAssembler.java:271)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:132)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4094)
	at org.eclipse.swt.widgets.Display.lambda$2(Display.java:3444)
	at org.eclipse.swt.widgets.Display.runWithProperDPIAwareness(Display.java:5415)
	at org.eclipse.swt.widgets.Display.msgFilterProc(Display.java:3442)
Caused by: java.lang.NumberFormatException: For input string: ""
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
	at java.base/java.lang.Integer.parseInt(Integer.java:672)
	at java.base/java.lang.Integer.parseInt(Integer.java:778)
	at java.base/java.text.MessageFormat.makeFormat(MessageFormat.java:1483)
	... 12 more
```

## Tested under this environment:
* OS & version: Windows 11
* Eclipse IDE/Platform version (as shown in *Help > About*): 2026-03 (4.39.0)

## Community

- [x] I understand reporting an issue to this OSS project does not mandate anyone to fix it. Other contributors may consider the issue, or not, at their own convenience. The most efficient way to get it fixed is that I fix it myself and contribute it back as a good quality patch to the project.

--- a/bundles/org.eclipse.e4.ui.workbench/META-INF/MANIFEST.MF
+++ b/bundles/org.eclipse.e4.ui.workbench/META-INF/MANIFEST.MF
@@ -1,7 +1,7 @@
 Manifest-Version: 1.0
 Bundle-ManifestVersion: 2
 Bundle-SymbolicName: org.eclipse.e4.ui.workbench;singleton:=true
-Bundle-Version: 1.18.200.qualifier
+Bundle-Version: 1.18.300.qualifier
 Bundle-Name: %pluginName
 Bundle-Vendor: %providerName
 Bundle-Localization: plugin

--- a/bundles/org.eclipse.e4.ui.workbench/src/org/eclipse/e4/ui/internal/workbench/ModelAssembler.java
+++ b/bundles/org.eclipse.e4.ui.workbench/src/org/eclipse/e4/ui/internal/workbench/ModelAssembler.java
@@ -39,6 +39,8 @@ import java.util.SortedSet;
 import java.util.TreeSet;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import org.eclipse.core.runtime.IConfigurationElement;
 import org.eclipse.core.runtime.IExtension;
 import org.eclipse.core.runtime.IExtensionPoint;
@@ -208,6 +210,8 @@ public class ModelAssembler {
 	private final AtomicReference<IExtensionRegistry> registry = new AtomicReference<>();
 
 	private final List<ServiceReference<IModelProcessorContribution>> processorContributions = new CopyOnWriteArrayList<>();
+
+	private static final Pattern MESSAGE_FORMAT_BRACKETS_PATTERN = Pattern.compile("\\{\\}"); //$NON-NLS-1$
 
 	private BundleContext bundleContext;
 
@@ -820,7 +824,7 @@ public class ModelAssembler {
 			logMethod.invoke(log, message, args);
 		} else {
 			// fallback if no LogService is available
-			stream.println(MessageFormat.format(message, args));
+			stream.println(MessageFormat.format(addIndex(message), args));
 		}
 	}
 
@@ -844,5 +848,21 @@ public class ModelAssembler {
 			this.factory = null;
 			this.logger = null;
 		}
+	}
+
+	/**
+	 * Adds missing indices to {@code {}} sequences in {@code format}. See:
+	 * {@link com.sun.org.slf4j.internal.Logger#addIndex(String)}
+	 */
+	private static String addIndex(String format) {
+		Matcher matcher = MESSAGE_FORMAT_BRACKETS_PATTERN.matcher(format);
+		StringBuilder sb = new StringBuilder();
+		int index = 0;
+		while (matcher.find()) {
+			String r = "{" + Integer.toString(index++) + "}"; //$NON-NLS-1$ //$NON-NLS-2$
+			matcher.appendReplacement(sb, r);
+		}
+		matcher.appendTail(sb);
+		return sb.toString();
 	}
 }

